app/services/foreman/provision/ssh.rb

```diff
@@ -50,64 +71,72 @@ def command_prefix
     (username == "root") ? "" : "sudo "
   end
 
-  def command
-    # Use the users home to store the provision script since we can't reliably
-    # tell if other locations are writeable or executable by the user.
+  # Use the users home to store the provision script since we can't reliably
+  # tell if other locations are writeable or executable by the user.
+  def main_command
     main_execution = "(chmod 0701 ./#{remote_script} && #{command_prefix} ./#{remote_script} ; echo $? >#{remote_script}.status) 2>&1"
-    "#{command_prefix} sh -c '#{main_execution} | tee #{remote_script}.log; exit $(cat #{remote_script}.status)'"
-  end
-
-  def defaults
-    {
-      :keys_only    => true,
-      :config       => false,
-      :auth_methods => %w(publickey),
-      :compression  => true,
-      :logger       => logger,
-    }
+    cleanup = "#{command_prefix} rm -f #{remote_script}"
+    "#{command_prefix} sh -c '#{main_execution} | tee #{remote_script}.log; #{cleanup}; exit $(cat #{remote_script}.status)'"
   end
 
   def logger
     Rails.logger
   end
 
-  def initiate_connection!
-    Timeout.timeout(Setting[:ssh_timeout].to_i) do
-      Timeout.timeout(8) do
-        ssh.run('pwd')
-      end
-    rescue Errno::ECONNREFUSED
-      logger.debug "Connection refused for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Errno::EHOSTUNREACH
-      logger.debug "Host unreachable for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::Disconnect
-      logger.debug "Host dropping connections for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::ConnectionTimeout
-      logger.debug "Host timed out for #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Net::SSH::AuthenticationFailed
-      logger.debug "Auth failed for #{username} at #{address}, retrying"
-      sleep(2)
-      retry
-    rescue Timeout::Error
-      retry
-    rescue => e
-      Foreman::Logging.exception("SSH error", e)
-    end
+  def scp
+    logger.debug "about to upload #{template} to remote system at #{remote_script}"
+    logger.debug scp_cmd
+    _stdout, stderr, status = Open3.capture3(*scp_cmd)
+
+    return if status.success?
+
+    logger.error stderr
+    raise Foreman::Exception.new("Failed to upload '#{@template}' to remote system at #{address}")
   end
 
-  def ssh
-    Fog::SSH.new(address, username, options.merge(:timeout => 4))
+  def passwd_cmd
+    return if @options[:password].empty?
+    ["sshpass", "-p", @options[:password]]
   end
 
-  def scp
-    Fog::SCP.new(address, username, options)
+  def ssh_key
+    return if @options[:key_data].empty?
+    ["-i", @tmp_key_path]
+  end
+
+  def ssh_cmd(host_command)
+    [passwd_cmd, "ssh", ssh_key, ssh_options, remote_host, host_command].flatten.compact
```

Comment:
This is my personal preference, but why is it needed to `flatten` and `compact` it? Do you expect `host_command`? I'd prefer to be explicit about what returns arrays. For example:
```suggestion
    passwd_cmd + ["ssh", ssh_key] + ssh_options + [remote_host, host_command]
```

And generally prefer returning arrays with `nil` entries so you don't need to compact it. It would also mean you return `[]` in `passwd_cmd` rather than `nil`.

Reply:
I need to return an array for the `Open3.capture3`, how would the change to `passwd_cmd + ["ssh", ssh_key] + ssh_options + [remote_host, host_command]` work with it?

I need `flatten` because `ssh_key`, `passwd_cmd`, and `ssh_options` are return arrays. Yes, these methods could return a string directly, but it was easier for me to read to have all methods return arrays.

Using `compact`
`ssh_key` or `passwd_cmd` can return `nil` if the condition is not met. Passing an array with `nil` option to  `Open3.capture3` causes a failure: 
```
TypeError: no implicit conversion of nil into String
```

I could have added a condition for adding these commands to the array in the `ssh_cmd` method itself, but I found using the `compact` easier to read.

Reply:
In case it's not needed you can return an empty array instead of `nil`. I always like to think about the exact data type it returns and right now (using Yard's type syntax, see https://yardoc.org/types.html) it's `Array<String>, nil` and because of that you need `.flatten.compact`. If you change it to `Array<String>` you can avoid the additional function call.

As an example, this is what I mean
```ruby
# @return [Array<String>] The password command
def passwd_cmd(mask_passwd: false)
  if @options[:password].empty?
    []
  else
    value = mask_passwd ? '*****' : @options[:password]
    ["sshpass", "-p", value]
  end
end

# @return [Array<String>] The options to add for the SSH key
def ssh_key
  @options[:key_data].empty? ? [] : ["-i", @tmp_key_path]
end

# @param [String] The command to run on the remote host
# @return [Array<String>] The SSH command to run
def ssh_cmd(host_command)
  passwd_cmd + ["ssh"] + ssh_key + ssh_options + [remote_host, host_command]
end
```

Reading this I now also think `ssh_key` should be as part of `ssh_options`. I also noticed it now doesn't pass `mask_passwd` so `ssh_cmd`.